frontend/src/Onboarding/TransferCoursesScreen.tsx

```diff
@@ -0,0 +1,363 @@
+ * @param reqs
+ */
+function flatten(reqs: Requirement[]): IRequiredCourse[][] {
+  return reqs.map(flattenOne).reduce((array, cur) => array.concat(cur), []);
+}
+
+function flattenOne(req: Requirement): IRequiredCourse[][] {
+  if (req.type === "COURSE") {
+    return [[req as IRequiredCourse]];
+  } else if (
+    req.type === "AND" &&
+    req.courses.filter(c => c.type === "COURSE").length
+  ) {
+    return [req.courses as IRequiredCourse[]];
+  } else if (req.type === "AND" || req.type === "OR") {
+    return flatten(req.courses);
+  } else {
+    return [];
+  }
+}
+
+interface TransferCoursesScreenProps {
+  major: Major;
+  completedRequirements: IRequiredCourse[];
+  setCompletedCourses: (completedCourses: ScheduleCourse[]) => void;
+  setTransferCourses: (transferCourses: ScheduleCourse[]) => void;
+}
+
+type Props = TransferCoursesScreenProps & RouteComponentProps;
+
+interface State {
+  selectedRequirements: IRequiredCourse[];
+  completedNonTransfer: IRequiredCourse[];
+  modalVisible: boolean;
+  otherCourses: IRequiredCourse[][];
+}
+
+class TransferCoursesComponent extends Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      modalVisible: false,
+      selectedRequirements: [],
+      otherCourses: [],
+      completedNonTransfer: props.completedRequirements,
+    };
+  }
+
+  hideModal() {
+    this.setState({ modalVisible: false });
+  }
+
+  showModal() {
+    this.setState({ modalVisible: true });
+  }
+
+  // Adds the "other courses" to the state in the form of a IRequiredCourse[][] so that they can be
+  // processed by renderAllCourses
+  addOtherCourses(courses: ScheduleCourse[]) {
+    let reqCourseMap = courses.map((course: ScheduleCourse) => [
+      {
+        type: "COURSE",
+        classId: +course.classId,
+        subject: course.subject,
+      } as IRequiredCourse,
+    ]);
+    this.setState({
+      otherCourses: [...this.state.otherCourses, ...reqCourseMap],
+    });
+  }
+
+  /**
+   * Separate transfer and non transfer courses, convert non transfer completed courses
+   * into ScheduleCourses, and populate redux store.
+   */
+  async onSubmit() {
+    let completedNonTransfer = this.state.completedNonTransfer.filter(
+      req => !this.state.selectedRequirements.includes(req)
+    );
+    let completed: ScheduleCourse[] = [];
+
+    for (let course of completedNonTransfer) {
+      const scheduleCourse = await fetchCourse(
+        course.subject,
+        String(course.classId)
+      );
+
+      if (scheduleCourse) {
+        completed.push(scheduleCourse);
+      }
+    }
+
+    let selectedCourses: ScheduleCourse[] = [];
+    for (let course of this.state.selectedRequirements) {
+      const scheduleCourse = await fetchCourse(
+        course.subject,
+        String(course.classId)
+      );
+
+      if (scheduleCourse) {
+        selectedCourses.push(scheduleCourse);
+      }
+    }
+
+    this.props.setCompletedCourses(completed);
```

Comment:
Just curious, why are completed courses being set here? It appears that we are taking the set completed requirements, filtering out those that are also taken as transfer credit, and re-setting the completed courses to that having already completed the completed courses step.

Is this to get rid of any overlap between what was selected as completed on the previous page and what was selected as transfer on this page?

Reply:
Yeah it's to distinguish completed and trasfer credits, since completed courses populate into the schedule and transfer credits go into the transfer credit modal (not yet but soon)